ArduCopter/takeoff_check.cpp

```diff
@@ -8,49 +8,19 @@
 void Copter::takeoff_check()
 {
 #if HAL_WITH_ESC_TELEM && FRAME_CONFIG != HELI_FRAME
-    // If takeoff check is disabled or vehicle is armed and flying then clear block and return
-    if ((g2.takeoff_rpm_min <= 0) || (motors->armed() && !ap.land_complete)) {
+    // If vehicle is armed and flying then clear block and return
+    if (motors->armed() && !ap.land_complete) {
         motors->set_spoolup_block(false);
         return;
     }
 
-    // block takeoff when disarmed but do not display warnings
-    if (!motors->armed()) {
-        motors->set_spoolup_block(true);
-        takeoff_check_warning_ms = 0;
-        return;
-    }
-
     // if motors have become unblocked return immediately
     // this ensures the motors can only be blocked immediate after arming
-    if (!motors->get_spoolup_block()) {
-        return;
-    }
-
-    // check ESCs are sending RPM at expected level
-    uint32_t motor_mask = motors->get_motor_mask();
-    const bool telem_active = AP::esc_telem().is_telemetry_active(motor_mask);
-    const bool rpm_adequate = AP::esc_telem().are_motors_running(motor_mask, g2.takeoff_rpm_min, g2.takeoff_rpm_max);
-
-    // if RPM is at the expected level clear block
-    if (telem_active && rpm_adequate) {
-        motors->set_spoolup_block(false);
+    if (motors->armed() && !motors->get_spoolup_block()) {
         return;
     }
 
-    // warn user telem inactive or rpm is inadequate every 5 seconds
-    uint32_t now_ms = AP_HAL::millis();
-    if (takeoff_check_warning_ms == 0) {
-        takeoff_check_warning_ms = now_ms;
-    }
-    if (now_ms - takeoff_check_warning_ms > 5000) {
-        takeoff_check_warning_ms = now_ms;
-        const char* prefix_str = "Takeoff blocked:";
-        if (!telem_active) {
-            gcs().send_text(MAV_SEVERITY_CRITICAL, "%s waiting for ESC RPM", prefix_str);
-        } else if (!rpm_adequate) {
-            gcs().send_text(MAV_SEVERITY_CRITICAL, "%s ESC RPM out of range", prefix_str);
-        }
-    }
+    // Run the common takeoff checks
+    motors->set_spoolup_block(!motors_takeoff_check(g2.takeoff_rpm_min, g2.takeoff_rpm_max));
```

Comment:
I am sure this is the right kind of refactoring, but I am having hard time convincing myself that before and after is equivalent for copter. Will need careful testing

Reply:
The hard line on `set_spoolup_block` and not blocking manual takeoffs in plane made it tough to split out the common code in a more-obviously-equivalent way. The order of some of these if-checks has shuffled for sure, which is tough.

If you have any loose suggestions for additional edge-cases to cover in the autotests, I can add them.

Reply:
I think just relying on the autotests and making sure this is flown on copter (where it is used a lot) is sufficient. @rmackay9 should also review since he was the driving force behind most of the copter changes